discovery-provider/src/models/tracks/track.py

```diff
@@ -24,40 +25,52 @@
 class Track(Base, RepresentableMixin):
     __tablename__ = "tracks"
 
-    blockhash = Column(String, ForeignKey("blocks.blockhash"), nullable=True)
-    blocknumber = Column(Integer, ForeignKey("blocks.number"), nullable=True)
-    slot = Column(Integer, nullable=True)
-    txhash = Column(String, default="", nullable=False)
-    track_id = Column(Integer, nullable=False)
-    is_current = Column(Boolean, nullable=False)
+    blockhash = Column(ForeignKey("blocks.blockhash"))  # type: ignore
+    track_id = Column(Integer, primary_key=True, nullable=False)
```

Comment:
For my own understanding, do the primary keys get ordered by declaration order here?

Reply:
Really glad you called this out. The answer is no, unfortunately. They're in the order the columns were added to the db, and we changed our PKey w/ the addition of txhash.

I did some digging, TL;DR, i don't think the codegen handles this particularly well (https://github.com/agronholm/sqlacodegen/issues/199#issuecomment-1114157530) but there is work to make it better. I think we should leave as is.

Since we don't tend to create tables via SQLAlchemy models, which i think is generally a bad bad idea, it doesn't have huge material impact.

We could add the PrimaryKeyConstraint lines by hand which are what respects the order in table creation, but I also don't love the idea of doing that by hand, which could lie.

There's also some differences between sql dialects in what is respected here https://stackoverflow.com/questions/51407143/sql-alchemy-composite-key-order

Reply:
I think we always should look at the DB, especially b/c python is not the only consumer. Best case scenario is the sqlacodegen thing could call this out about the primary keys, but i suppose it can't.

I think if we have sqlalchemy generate models, we'll still have to check the db to make sure that the models are "doing the right thing." time and again, I think we've just seen sqlalchemy do somewhat nonsensical things. Our migrations that do things as raw sql tend to go way nicer.

Do you have a suggestion to change something in this PR that would make you feel better?

Reply:
I don't have any suggestions/change requests on the PR wrt this, more-so trying to understand from a philosophical perspective how we can avoid having large PRs to reconcile the model and DB state going forward.

It seems to me like having separate migrations without updating models is how we ended up in this state to begin with, so my theory is that it's more trustworthy to go "model-first" approach since at least that more declaratively states intent and ensures models represent the desired state, and then touch up the migrations if the autogenerated ones don't represent it appropriately. I'm curious what issues we run into with alembic? I'm aware of some wrt idempotent migrations, but it's not getting column types wrong or anything, right? I want to be clear I'm not opposed to SQL migrations as long as models get updated appropriately, which I think is more likely if we update models as a first step and generate migrations from there. Even if we don't use the autogenerated migration, it will let us know what the model/ORM thinks the DB state is which can inform the code for the SQL migration.

Alternatively we could do "db-first" and then autogenerate models using the approach you did for this PR, but that seems to have caveats too, such as here with the PrimaryKeyConstraint, or if the migration you're writing and testing fails to clean something up from a previous test run, or maybe not all tables should be modeled per your comment about not only python, etc. It also requires more of a manual lift/comparison against our whole model repertoire I assume. I'm guessing it's not as easily baked into the process as the autogeneration for migrations is, and so I'm wary that the extra friction to the process of constantly running that will make it less likely to be followed. Maybe I'm wrong here and there's an easy way to diff the models after coding a migration using the tool you used for that without needing any extra work?

In this specific case regarding primary keys, it came up once before that we had slow queries because `is_current` wasn't included in a query so a query couldn't use the PKEY since it was the first part of the composite. We've also had issues before where we didn't have indexes where we expected them. I think these kinds of issues are perhaps not solved but more easily identified if we can trust models more, whether that means keeping models in sync with the db or the db in sync with models as closely as possible.

tl;dr: models should tell as much truth as possible imo, and the best way to achieve that is a "model-first" approach to DB design (excuse my misuse of EntityFramework's "model-first" - in their nomenclature, it's actually "code-first")

Reply:
Resolved offline - keeping existing workflow for now, will revisit if seeing more divergence